### ISSUE TITLE
test_pthread_run_on_main_thread_flood fails silently

### ISSUE THREAD

**Issue by sbc100** [emscripten-core/emscripten#12885]
The thread return code from pthread_join is not checked so when an assert fires in the thread the main thread does not record the failure.

**Comment by sbc100** [comment]
Uncommenting the assert in #12886 has this test failing 1/N times where N seems to be between 5 and 10. 

The final_count in those cases is around 930 instead of 1000.

Any ideas what could be going on here?

**Comment by sbc100** [comment]
When investing another issue with this test I noticed that it seem to be using the `emscripten_async_waitable` in way that the API doesn't support.

This is causing UAF and intermittent failures in another PR I'm working on that merges libpthread into libc.   It seems that building `library_pthread.c` with `-Os` (libc setting) rather than `-O2` the current setting libpthread makes this bug more visible.

If you take a look at `test_async_waitable` it tries to run a bunch of async tasks using `emscripten_async_waitable_run_in_main_runtime_thread` but frees them right away rather than waiting on them.   The API currently doesn't seem to support calling `emscripten_async_waitable_close` until the wait has been done.   

Is that supposed to work of is the test just wrong?
